### ISSUE TITLE
Ungraceful handling of solver exit

### ISSUE THREAD

**Issue by robdockins** [GaloisInc/crucible#322]
The s-expression parser for solver responses seems to assume that a responses are newline separated, and that a single response will occur on a single line.  This is generally true, but some solvers (in particular CVC4) will sometimes return error reponses where newlines appear directly in string constants.  These are incorrectly parsed and instead of seeing an informative error message we get:

```
Exception: Could not parse solver response:
      Could not parse acknowledgement result.
    *** Exception: Parse exception: Failed reading: empty

    in response to commands:
      (pop 1)
```

The raw solver interaction output reveals the true error:
```
(set-option :print-success true)
; success
(set-option :produce-models true)
; success
(set-logic ALL_SUPPORTED)
; success
(push 1)
; success
; :1:0
(declare-datatypes () ((Struct0 mk-struct0)))
; (error "Parse Error: <stdin>:5.24: Too many datatypes defined in this block.
;
;   (declare-datatypes () ((Struct0 mk-struct0)))
;                           ^
; ")
(pop 1)
```

**Comment by robdockins** [comment]
Turns out, I misdiagnosed this problem.  The multiline error is actually being read correctly, and the parsing system correctly pulls in characters as necessary independent of newlines.

Instead, what is happening is that CVC4 is immediately exiting after the error instead of remaining online and responding to further requests.  As a result, the subsequent `(pop)` command results in an error when we try to read the response, because the pipe is closed.  This masks the original error and we instead see a "broken pipe" message.

Ideally, we can find some way to put CVC4 into a mode where it continues after errors.  Failing that, we need to somehow deal with the fact that the solver may die following an error.
